sequencer/src/block2.rs

```diff
@@ -244,850 +71,10 @@ impl BlockPayload for Payload {
     }
 
     fn encode(&self) -> Result<Self::Encode<'_>, Self::Error> {
-        Ok(self.payload.iter().cloned())
+        Ok(self.raw_payload.iter().cloned())
```

Comment:
It seems that changing the signature of the `encode` function will have impact in a lot of places, including HotShot.
https://github.com/EspressoSystems/HotShot/blob/171e22a0c6bf53cd127adb6ce12ad5e58553aefc/crates/hotshot/src/lib.rs#L209

Reply:
This should have no impact downstream because downstream code *should* talk to the payload only via `BlockPayload` trait. That trait promises only `Encode<'a>: 'a + Iterator<Item = u8> + Send`; the proposed change does not violate this promise.

Admittedly, the downstream code you link merely collects & copies the iterator anyway, so this specific downstream use would not benefit from the proposed change. But I do think it's good practice to do it anyway so that hotshot can unilaterally take advantage of this optimization at any time in the future. Indeed, there's even a TODO for it in the hotshot code you cite that links to https://github.com/EspressoSystems/HotShot/issues/2115.

That said, I don't think it's possible to get a `Iterator<Item = u8>` without cloning. Seems that a proper solution requires modifying the `BlockPayload` trait so that the assoc type `Encode` is able to support non-copying of the payload downstream. (eg. `Encode: AsRef<[u8]>` or something.) But that's a significant change so we should not touch it in this PR.

Instead, I suggest a code comment like
```rust
// TODO change `BlockPayload::Encode` trait bounds to enable copyless encoding such as AsRef<[u8]>
// https://github.com/EspressoSystems/HotShot/issues/2115
```

Reply:
Yes, comment added in e7ccdbc.